modules/frontend/dedicated_columns_to_json.go

```diff
@@ -0,0 +1,44 @@
+package frontend
+
+import (
+	"encoding/json"
+	"unsafe"
+
+	"github.com/grafana/tempo/tempodb/backend"
+)
+
+type DedicatedColumnsToJSON struct {
+	columnsToJSON map[uint64]string
+}
+
+func newDedicatedColumnsToJSON() *DedicatedColumnsToJSON {
+	return &DedicatedColumnsToJSON{
+		columnsToJSON: make(map[uint64]string),
+	}
+}
+
+func (d *DedicatedColumnsToJSON) JSONForDedicatedColumns(cols backend.DedicatedColumns) (string, error) {
+	if len(cols) == 0 {
+		return "", nil
+	}
+
+	hash := cols.Hash()
+	if jsonString, ok := d.columnsToJSON[hash]; ok {
+		return jsonString, nil
+	}
+
+	proto, err := cols.ToTempopb()
+	if err != nil {
+		return "", err
+	}
+
+	jsonBytes, err := json.Marshal(proto)
+	if err != nil {
+		return "", err
+	}
+
+	jsonString := unsafe.String(unsafe.SliceData(jsonBytes), len(jsonBytes))
```

Comment:
```go
jsonString := string(jsonBytes)
```

Reply:
this will incur another allocation/clone of the byte slice. given that we have full control over the lifecycle of the byte slice it feels safe.